docs(docs-readme): fill out docs-readme output target (#1332)
ionic-team/stencil-site: docs/documentation-generation/docs-readme.md

# Docs Readme Markdown File Auto-Generation

Stencil is able to auto-generate `readme.md` files for your components.
This can help you to maintain consistently formatted documentation for your components which lives right next to them and render in GitHub.

## Setup

To generate markdown files, it is recommended to add the `docs-readme` output target to your Stencil configuration file:

```tsx title="stencil.config.ts"
import { Config } from '@stencil/core';

export const config: Config = {
  outputTargets: [
    { 
      type: 'docs-readme'
    }
  ]
};
```

## Generating README Files

### Using the Build Task

> If your project has a `docs-readme` output target configured in your Stencil configuration file, the Stencil [build task](../config/cli.md#stencil-build) is all that's needed to generate README docs:

Author: 👍 updated in ef12b550c

```bash
npx stencil build
```
If you're running the build task with the `--watch` flag, your project's README files will automatically update without requiring multiple explicit build commands:
```bash
npm stencil build --watch
```

:::info
When running the build task with the `--dev` flag, README files will not be generated.
This is to prevent unnecessary I/O operations during the development cycle.
:::

If you choose not to include a `docs-readme` output target in your Stencil configuration file, use the `--docs` CLI flag as a part of the build task:
```bash
npx stencil build --docs
```

This will cause the Stencil compiler to perform a one-time build of your entire project, including README files.

> ### Using the Docs Task

Author: 👍 updated in ef12b550c - I went with 'Command' here

As an alternative to the build task, the [docs task](../config/cli.md#stencil-docs) can be used to perform a one time generation of the documentation:
```bash
npx stencil docs
```
Running `stencil docs` will generate documentation for [all documentation output targets](./01-overview.md), not just `docs-readme`.

## README Sections

Most generated markdown content will automatically be generated without requiring any additional configuration.
Content is generated based on its Stencil component, rather than requiring you to configure multiple flags.
Each section below describes the different types of content Stencil recognizes and will automatically generate. 

### Custom Markdown Content

Once you've generated a `readme.md` file, you can add your own markdown content to the file.
You may add any content above the following comment in a component's `readme.md`:
```
Custom content goes here!
<!-- Auto Generated Below -->
```

Any custom content placed above this comment will be persisted on subsequent builds of the README file.

### Deprecation Notices

A Stencil component may be marked as deprecated using the [JSDoc `@deprecated` tag](https://jsdoc.app/tags-deprecated).

> By placing `@deprecated` in a component's class-level JSDoc will cause the generated README to denote the component is deprecated.

Reviewer: ```suggestion
By placing `@deprecated` in a component's class-level JSDoc it will cause the generated README to denote the component as deprecated.
```

In this case, that description is "since v2.0.0".

The deprecation notice will be placed after the [custom content](#custom-markdown-content) in the README.

If a component is not marked as deprecated, this section will be omitted from the generated README.

### Component Overview

A Stencil component that has a JSDoc comment on its class component like so:

```tsx title="A component with an overview in its JSDoc"
/**
 * A simple component for formatting names
 *
 * This component will do some neat things!
 */
@Component({
  tag: 'my-component',
  shadow: true,
})
export class MyComponent { }
```
will generate the following section in your component's README:

```
## Overview

A simple component for formatting names

This component will do some neat things!
```

The overview will be placed after the [deprecation notice](#deprecation-notices) section of the README.

If a component's JSDoc does not contain an overview, this section will be omitted from the generated README.

### Usage Examples

Usage examples are user-generated markdown files that demonstrate how another developer might use a component.

> These files are separate from a component's README file, and are placed in a `usage/` adjacent to a component's implementation.

Reviewer: ```suggestion
These files are separate from a component's README file, and are placed in a `usage/` directory adjacent to a component's implementation.
```

   * @param duration The amount of time to take scrolling by that amount.
   * @returns the total distance travelled
   */
  @Method()
  async scrollByPoint(x: number, y: number, duration: number): Promise<number> { /* omitted */ }

  // ...
}
```

The following section will be generated:

```md
## Methods

### `scrollByPoint(x: number, y: number, duration: number) => Promise<number>`

Scroll by a specified X/Y distance in the component.

#### Parameters

| Name       | Type     | Description                                          |
| ---------- | -------- | ---------------------------------------------------- |
| `x`        | `number` | The amount to scroll by on the horizontal axis.      |
| `y`        | `number` | The amount to scroll by on the vertical axis.        |
| `duration` | `number` | The amount of time to take scrolling by that amount. |

#### Returns

Type: `Promise<number>`

the total distance travelled
```

The methods section will be placed after the [@Event section](#event-details) of the README.

If a component does not use the `@Method()` decorator, this section will be omitted from the generated README.

### @slot Details

> A component that uses [slots](../components/templating-and-jsx.md#slots) may describe the its slots in the component's JSDoc using the Stencil-specific `@slot` JSDoc tag.

Reviewer: ```suggestion
A component that uses [slots](../components/templating-and-jsx.md#slots) may describe its slots in the component's JSDoc using the Stencil-specific `@slot` JSDoc tag.
```
